### PR TITLE
Dev amc

### PR DESCRIPTION
**Pull request by chrisruk**
Hey, 

Just submitting a pull req. to merge to master.

If theres any issues give me a shout!

Cheers
Chris

### CODE REVIEW

File: grc/CMakeLists.txt [hunk 1]
```diff
@@ -23,6 +23,8 @@ install(FILES
     inspector_qtgui_sink_vf.xml
     inspector_block_tree.xml
     inspector_ofdm_zkf_c.xml
+    inspector_sync_prototype_cc.xml
```
Comment (sbmueller) [inline grc/CMakeLists.txt]: @chrisruk This prototype xml is deprecated

Comment (chrisruk) [reply grc/CMakeLists.txt]: Just removed that, cheers.

**Comment by sbmueller** [comment]
Thanks for bringing this up. I have had a first look at your branch and noticed some things, that should be done/revisited before merging:

1) When CMaking, there is still this warning:

```
CMake Warning (dev) at CMakeLists.txt:193 (link_directories):
  This command specifies the relative path

    QWTPLOT3D_LIBRARIES

  as a link directory.

  Policy CMP0015 is not set: link_directories() treats paths relative to the
  source dir.  Run "cmake --help-policy CMP0015" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.
This warning is for project developers.  Use -Wno-dev to suppress it.
```

It would be nice if you could set the policy in the CMakeLists.txt to avoid this (there are already some other policies where you can have a look).

2) You look for OpenGL in the GrSetupQt4.cmake. This seems to work for both our machines, but having a look at a much more extensive template https://github.com/Kitware/CMake/blob/master/Modules/FindOpenGL.cmake makes me think if there aren't other possibilities, where this doesn't work. Can you revisit this? I can also have a look on OS X in the next days.

3) You should improve Doxygen documentation. Beginning at the main page where your blocks should be mentioned and install instructions should be updated. Also, your blocks have very little documentation text. For me it is unfortenately not clear how your blocks work after reading it. Please make sure to name all parameters and add descriptions for all parameters. You can look at my blocks for the corresponding Doxygen keywords.
Also please make sure to only document in the public header in `include/inspector`. This is GR standard (else the documentation will show up twice in GRC as I noticed).

4) Please add yourself as maintainer in the README.md and Doxygen main page if you intend to do so.

5) I would appreciate if your blocks could get a more descriptive title. I'm not a deep learning person and I can not really tell what `tfmodel` should mean. Also imagine having a flowgraph where the context is not clear (in opposite to the tree view). It should be made clear what's the purpose of the block, for instance "Tensorflow AMC Model". Same for vis3d (maybe "3D Visualizer Block"?).
Also, please make sure to rename the classes correctly following the GR coding guidelines: http://gnuradio.org/redmine/projects/gnuradio/wiki/Coding_guide_impl#Block-naming-conventions.

6) As I assume you don't ship a TF model now within the repo. This is fine, but there should be instructions on how to generate one locally. Please add a separate page for that in Doxygen. I remember some issues generating it (Keras config etc.)

7) I had a look at the examples and all seem to work, nice! Still the examples do not really make clear to me how to use the blocks. There are for instance Vector_to_Stream(2*Np) -> Stream_to_Vector(2 \* P \* L \* 2 \* Np) blocks chained. Why is that? What do these parameters mean? It should be made clear for an user without any foreknowledge about the toolbox how to use it.
Can you also include the Inspector GUI in at least one example, to make clear that feedback can be viewed in the GUI?

8) Please add your blocks to `grc/inspector_block_tree.xml`.

9) I noticed you use string type for the model file path in your XMLs. You should rather use `file_open`, which allows the user to open a file choosing dialogue.

10) I've seen you added a cpp & python unit test for vis3d and a python one for tfmodel. It seems all of them are empty, or have I missed something? AFAIK vis3d is only a GUI block, so you can remove the cpp test here since it has no function. Also please add a proper unit test code for tfmodel.

I'll have a look at the code changes you made and comment on that if there is anything left. Nice work, @chrisruk , really looking forward to merge these blocks into master :)

**Comment by chrisruk** [comment]
@sbmueller Just been going through your comments:

1) I've set the policy, the warning seems to have gone for me
2) I'll look into this
3) I've been working on improving the doxygen generation, for the C++ vis3d block, I can't seem to work out why the documentation is still duplicated though, any ideas would be most appreciated! Also, can python blocks be documented in a way, so that their documentation appears in GRC?
4) Done
5) Renamed the blocks to conform with GNU Radio naming and renamed for GRC
6) Added documentation in doxgen for the installation. 
7) Added a note about why stream to vector is used, in the GRC flow graph, will add some more info too, I'll add another GRC flow graph combining inspector
8) Done
9) Done
10) Not sure how the TFmodel can be tested at the moment without a model file - will have a think

Cheers!

**Comment by sbmueller** [comment]
11) Change copyright according to Martin's mail.

Thanks for the quick changes. I'll be away over the weekend, so take your time. I'll review & merge next week.

**Comment by sbmueller** [comment]
About 3): The GRC documentation text is just the docstring of the python class. If you add a docstring there, it should appear in GRC.
Please ping me when everything is ready, then I'll merge into master.

**Comment by chrisruk** [comment]
I've modified the doxygen file so that Python documentation is created but can't seem to get it to appear in GRC for some reason, I tried adding a comment using the triple quote marks too, but that didn't seem to make a difference.

Also do you have any idea why the documentation for the TFModel block is being duplicated, at the moment I'm getting, the following in GRC:

```
vis3d_vf
make(int vlen, int width, int height, int gwidth, int gheight, double maxz, int fps, char * xaxis, char * yaxis, char * zaxis, QWidget * parent=None) -> vis3d_vf_sptr
        3D visualisation for the inspector toolkit
        This GUI uses QwtPlot3D to visualise data in 3D, for instance data from the gr-specest FAM block.
        Controls: Hold left mouse button and move cursor to rotate. Dragging using two fingers, zooms in or out
        Constructor Specific Documentation:
        Build a 3D visualisation.
        Args:
            vlen : The size of the input vector containing the data to be displayed
            width : The 1D vector will be reshaped with this width
            height : The 1D vector will be reshaped with this height
            gwidth : The graph will be stretched to this width
            gheight : The graph will be stretched to this height
            maxz : You can set the graph to have a maximum Z value, so that the graph boundaries remain fixed, or if you set to 0, the Z axis of the graph, will grow according to the data.
            fps : You can set the framerate of the graph, so that if a higher number vectors are received than the frame rate, they can be discarded.
            xaxis : Label for X axis
            yaxis : Label for Y axis
            zaxis : Label for Z axis
            parent : QWidget parent

vis3d_vf_make
vis3d_vf_make(int vlen, int width, int height, int gwidth, int gheight, double maxz, int fps, char * xaxis, char * yaxis, char * zaxis, QWidget * parent=None) -> vis3d_vf_sptr
    3D visualisation for the inspector toolkit
    This GUI uses QwtPlot3D to visualise data in 3D, for instance data from the gr-specest FAM block.
    Controls: Hold left mouse button and move cursor to rotate. Dragging using two fingers, zooms in or out
    Constructor Specific Documentation:
    Build a 3D visualisation.
    Args:
        vlen : The size of the input vector containing the data to be displayed
        width : The 1D vector will be reshaped with this width
        height : The 1D vector will be reshaped with this height
        gwidth : The graph will be stretched to this width
        gheight : The graph will be stretched to this height
        maxz : You can set the graph to have a maximum Z value, so that the graph boundaries remain fixed, or if you set to 0, the Z axis of the graph, will grow according to the data.
        fps : You can set the framerate of the graph, so that if a higher number vectors are received than the frame rate, they can be discarded.
        xaxis : Label for X axis
        yaxis : Label for Y axis
        zaxis : Label for Z axis
        parent : QWidget parent
```

I was thinking about the TFModel block, I don't think it can actually be tested without a model, so I'll remove the test for that.

**Comment by sbmueller** [comment]
Ok, thats not that important. AFAIK no in-tree python block has any GRC documentation.

**Comment by sbmueller** [comment]
Thanks for all the new commits. Can you revise the following:

7) I couldn't find a note or explanation for the Stream To Vector parameters. Where can I find them? Also, in the flowgraph with the Inspector GUI, please feed back your block to the GUI instead of a Message Debug block. This flowgraph sould show the final merged product, so please no Debug blocks anymore.
8) I think you need to rename your blocks in the `block_tree.xml` (vf and vcf).
10) I can still see empty `qa_vis3d_vf.*` files in the lib folder. I think you intended to remove them?

The rest looks fine, looking forward to the final commits :)

**Comment by sbmueller** [comment]
Nice work, I'll merge now.

**Pull request merged**